docs/source/user/theory.rst

```diff
@@ -0,0 +1,890 @@
+
+Many tridiagonal algorithms face performance challenges, particularly along the :math:`x` direction, due to inefficient memory access patterns in Cartesian data structures. There are two main characteristics in tridiagonal matrix algorithms that are crucial for computational performance:
+
+1. The algorithms are heavily bandwidth-bound because the FLOP requirements of these algorithms are very low compared to the data movement requirements.
+2. The backward and forward passes of the algorithm require accessing the same data twice in a short time interval. This can be optimised by using CPU cache or GPU shared memory to store intermediate states between the forward and backward passes.
+
+x3d2's solution: DistD2-TDS
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To address the challenges of solving tridiagonal systems efficiently in distributed-memory environments on both CPUs and GPUs, x3d2 uses a novel algorithm called `DistD2-TDS` (see :cite:`akkurt_cpc_24` for more details). This algorithm is based on a specialised data structure that:
+
+1. Improves data locality and minimises data movements via cache blocking and kernel fusion strategies.
+2. Uses data continuity to enable a contiguous data access pattern, resulting in efficient utilisation of the available memory bandwidth.
+3. Supports a data layout that enables vectorisation on CPUs and thread-level parallelisation on GPUs for improved performance.
+
+Modern CPUs have large enough caches to store the entire tridiagonal matrix, which can be accessed quickly. However, GPU caches are not large enough, and kernel fusion is the only option to reduce data movement on GPUs. To enable a linear and predictive memory access pattern regardless of the spatial direction of the tridiagonal systems, DistD2-TDS data structure subdivides the computational domain into groups of individual tridiagonal systems. These groups are tightly packed in memory to ensure data continiuty. This arrangement enables vectorisation on CPUs and thread-level parallelism on GPUs, as the :math:`n^{th}` entries of all tridiagonal systems within a group are stored next to each other in memory. Therefore, the sequential operations in the algorithms, as we apply the forward and backward passes, can be concurrently executed for :math:`SZ` systems at once per core on a CPU or per SM on a GPU.
+
+Example: In a domain of size :math:`32 \times 8 \times 4` shown below; a single group might consist of 4 individual tridiagonal systems, resulting in 8 groups in total. For a Cartesian mesh with :math:`n_x`, :math:`n_y`, and :math:`n_z` the data is arranged as  :math:`SZ, n_x, n_y \cdot n_z / SZ` where :math:`SZ` is the size of the group.
```

Comment:
In the example... a single group ~~might~~ consist of 4 individual tridiagonal systems. (better not to use 'might')
SZ is the number of tridiagonal systems in a group. (we use size mostly for tridiagonal system size).
(SZ, nx, ny*nz/SZ) **for x-directional layout** where SZ is ...